File: pages/inventory_page.py
from playwright.sync_api import Page, expect

from pages.base_page import BasePage


class InventoryPage(BasePage):
    def __init__(self, page: Page):
        super().__init__(page)
        self.title = page.locator(".title")
        self.inventory_items = page.locator(".inventory_item")
        self.cart_badge = page.locator(".shopping_cart_badge")
        self.cart_link = page.locator(".shopping_cart_link")

    def expect_products_page(self) -> None:
        expect(self.title).to_have_text("Products")

    def add_product(self, product_name: str) -> None:
        self.inventory_items.filter(has_text=product_name).locator("button").click()

    def expect_cart_count(self, count: int) -> None:
        if count == 0:
            expect(self.cart_badge).to_have_count(0)
        else:
            expect(self.cart_badge).to_have_text(str(count))

    def open_cart(self) -> None:
        self.cart_link.click()

File: pages/inventory.page.ts
import { expect } from '@playwright/test';
import { BasePage } from './base.page';

export class InventoryPage extends BasePage {
  private readonly title = this.page.locator('.title');
  private readonly inventoryItems = this.page.locator('.inventory_item');
  private readonly cartBadge = this.page.locator('.shopping_cart_badge');
  private readonly cartLink = this.page.locator('.shopping_cart_link');

  async expectProductsPage() {
    await expect(this.title).toHaveText('Products');
  }

  async addProduct(productName: string) {
    const item = this.inventoryItems.filter({ hasText: productName });
    await item.locator('button').click();
  }

  async expectCartCount(count: number) {
    if (count === 0) {
      await expect(this.cartBadge).toHaveCount(0);
    } else {
      await expect(this.cartBadge).toHaveText(String(count));
    }
  }

  async openCart() {
    await this.cartLink.click();
  }
}
